### PR TITLE
[WIP] - Entities - Discussion base

### PR DESCRIPTION
**Pull request by nupplaphil**
This is a pure discussion Draft - PR, which shows how Entities **could** be used in Friendica.

I will add comments at places, where the "magic" happens and what I thought about it.

I don't think that we will use it in this way, but this would be how I would "see" entities in Friendica.

Ideas are based on:
- https://designpatternsphp.readthedocs.io/en/latest/More/Repository/README.html#
- https://designpatternsphp.readthedocs.io/en/latest/Creational/FactoryMethod/README.html
- https://designpatternsphp.readthedocs.io/en/latest/Creational/Prototype/README.html

### CODE REVIEW

File: src/Model/Entity/Api/Mastodon/Instance.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * @property-read string thumbnail
+ * @property-read array languages
+ * @property-read int max_toot_chars
+ * @property-read boolean registratrions
+ * @property-read boolean approval_required
+ * @property-read Account|null contact_account
+ */
+class Instance extends BaseEntity
+{
+	use DataTransferObjectTrait;
```
Comment (nupplaphil) [inline src/Model/Entity/Api/Mastodon/Instance.php]: Annotating `@property-read` in combination with this trait `DataTransferObjectTrait` will make it impossible to set values to the given properties

File: src/Model/Entity/BaseEntity.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+	/**
+	 * Private constructor of the entity
+	 * Entities are not meant to get created manually by hand
+	 *
+	 * Instead use the static create() function
+	 *
+	 * @param array $data
+	 */
+	private function __construct(array $data = [])
```
Comment (nupplaphil) [inline src/Model/Entity/BaseEntity.php]: making the constructor `private` means that it is impossible to create entities without using:
- `BaseEntity::create()`
- `BaseEntity::createFromPrototype()`

File: src/Model/Entity/BaseEntity.php [hunk 3]
```diff
@@ -0,0 +1,102 @@
+<?php
+
+namespace Friendica\Model\Entity;
+
+use ArrayAccess;
+use Friendica\Network\HTTPException;
+
+abstract class BaseEntity implements ArrayAccess
```
Comment (nupplaphil) [inline src/Model/Entity/BaseEntity.php]: This is the base class of every entities of Friendica.
Means it is used for:
- `Business objects`: Introduction, Item, Contact, ...
- `Data Transfer Objects`: Mastodon Instance, Mastodon Account

It has the very basic functionality, which each entity in Friendica must have.

File: src/Model/Entity/BaseEntity.php [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+		}
+
+		if (!array_key_exists($name, $this->data)) {
+			throw new HTTPException\InternalServerErrorException(sprintf("Field '%s' not found in %s", $name,  static::class));
+		}
+
+		return $this->data[$name];
+	}
+
+	public static function create(array $data)
```
Comment (nupplaphil) [inline src/Model/Entity/BaseEntity.php]: A plain creator, which can get overriden by specific entities, if they need some "more" magic.

File: src/Model/Entity/BaseEntity.php [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	{
+		return new static($data);
+	}
+
+	public static function createFromPrototype($prototype, array $data)
```
Comment (nupplaphil) [inline src/Model/Entity/BaseEntity.php]: This one is for speed. You pass a created entity to this method and "just" override the data behind it.
It is far more faster than plain instancing

File: src/Model/Entity/BussinessObject/Contact.php [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ * @property boolean  blocked
+ * @property boolean  hidden
+ * @property string   name
+ * @property string   nickname
+ * @property string   url
+ * @property int      relation
+ * @property string   protocol
+ * @property string   network
+ */
+class Contact extends BaseEntity implements IStorable
```
Comment (nupplaphil) [inline src/Model/Entity/BussinessObject/Contact.php]: `IStorable` forces this entity to be storable on our persistence layer

File: src/Model/Entity/BussinessObject/Contact.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+{
+	use StorableTrait;
```
Comment (nupplaphil) [inline src/Model/Entity/BussinessObject/Contact.php]: The "standard" implementation of a `IStorable` entity, means plain setting of new values and saving the changed keys for later.

File: src/Model/Entity/IStorable.php [hunk 8]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace Friendica\Model\Entity;
+
+/**
+ * @property int id
+ */
+interface IStorable
```
Comment (nupplaphil) [inline src/Model/Entity/IStorable.php]: This interface is necessary to use "abstract" Repository features (select, delete, save).
Each entity, which is persistent to the database must implement this interface.

File: src/Model/Entity/Traits/DataTransferObjectTrait.php [hunk 9]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace Friendica\Model\Entity\Traits;
+
+use Friendica\Network\HTTPException;
+
+trait DataTransferObjectTrait
+{
+	public function __set($name, $value)
```
Comment (nupplaphil) [inline src/Model/Entity/Traits/DataTransferObjectTrait.php]: This one catches situations, where we want to override Data Transfer Objects (= Api, Mastodon Instance, ...)

File: src/Model/Entity/Traits/StorableTrait.php [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+		return count($this->changed) > 0;
+	}
+
+	public function getChanged()
+	{
+		$this->changed = array_unique($this->changed);
+		return $this->changed;
+	}
+
+	public function isStored()
```
Comment (nupplaphil) [inline src/Model/Entity/Traits/StorableTrait.php]: If true, the entitiy was already saved to the database.

File: src/Model/Entity/Traits/StorableTrait.php [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+	 * @var array
+	 */
+	private $changed = [];
+
+	/**
+	 * True, if the current entity has changes
+	 *
+	 * @return boolean
+	 */
+	public function isChanged()
```
Comment (nupplaphil) [inline src/Model/Entity/Traits/StorableTrait.php]: If true, the current entity was changed and needs to be saved

File: src/Model/Entity/Traits/StorableTrait.php [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+	{
+		return count($this->changed) > 0;
+	}
+
+	public function getChanged()
+	{
+		$this->changed = array_unique($this->changed);
+		return $this->changed;
```
Comment (nupplaphil) [inline src/Model/Entity/Traits/StorableTrait.php]: returns the changed keys (for optimized saving to the db)

File: src/Model/Factory/MastodonFactory.php [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+use Friendica\App\BaseURL;
+use Friendica\Core\Config\Configuration;
+use Friendica\Database\Database;
+use Friendica\Model\APContact;
+use Friendica\Model\Entity\Api\Mastodon\Instance;
+use Friendica\Model\User;
+use Friendica\Module\Register;
+use Psr\Log\LoggerInterface;
+
+class MastodonFactory extends BaseFactory
```
Comment (nupplaphil) [inline src/Model/Factory/MastodonFactory.php]: This one is a Factory, means it creates entities, not more!

File: src/Model/Factory/MastodonFactory.php [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+		$this->dba     = $dba;
+	}
+
+	/**
+	 * @return Instance
+	 *
+	 * @throws \Friendica\Network\HTTPException\InternalServerErrorException
+	 * @throws \ImagickException
+	 */
+	public function createInstance()
```
Comment (nupplaphil) [inline src/Model/Factory/MastodonFactory.php]: This method creates a Mastodon instance entity, which uses the "DataTransferObjectTrait". So the returned instance is not possible to be changed.

File: src/Model/Factory/MastodonFactory.php [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+			$adminList     = explode(',', str_replace(' ', '', $this->config->get('config', 'admin_email')));
+			$administrator = User::getByEmail($adminList[0], ['nickname']);
+			if (!empty($administrator)) {
+				$adminContact    = $this->dba->selectFirst('contact', [], ['nick' => $administrator['nickname'], 'self' => true]);
+				$apContact       = APContact::getByURL($adminContact['url'], false);
+				$contact_account = Account::create($this->baseUrl, $adminContact, $apContact);
+			}
+		}
+
+		return self::$entity::create([
```
Comment (nupplaphil) [inline src/Model/Factory/MastodonFactory.php]: Uses the creation-method of the current entity, this time the Mastodon instance :-)

File: src/Model/Factory/MastodonFactory.php [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+	/** @var Database */
+	protected $dba;
+
+	/**
+	 * @param LoggerInterface $logger
+	 * @param BaseURL         $baseUrl
+	 * @param Configuration   $config
+	 * @param Database        $dba
+	 */
+	public function __construct(LoggerInterface $logger, BaseURL $baseUrl, Configuration $config, Database $dba)
```
Comment (nupplaphil) [inline src/Model/Factory/MastodonFactory.php]: injects all dependencies, necessary for creating an Instance for the Mastodon API

File: src/Model/Repository/BaseRepository.php [hunk 17]
```diff
@@ -0,0 +1,116 @@
+<?php
+
+
+namespace Friendica\Model\Repository;
+
+use Friendica\Database\Database;
+use Friendica\Model\Entity\BaseEntity;
+use Friendica\Model\Entity\IStorable;
+
+abstract class BaseRepository
```
Comment (nupplaphil) [inline src/Model/Repository/BaseRepository.php]: This is the base of each repository. Every repository, which interacts with a table in Friendica has to extend this class

File: src/Model/Repository/BaseRepository.php [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+{
+	/** @var string */
+	protected static $table;
```
Comment (nupplaphil) [inline src/Model/Repository/BaseRepository.php]: sets the table behind the repository

File: src/Model/Repository/BaseRepository.php [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+	/** @var BaseEntity */
+	protected static $entity;
```
Comment (nupplaphil) [inline src/Model/Repository/BaseRepository.php]: sets the entity class, which is returned from this repository

File: src/Model/Repository/BaseRepository.php [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+	abstract protected function createEntity(array $data);
+
+	/**
+	 * @param array $conditions
+	 *
+	 * @return mixed
+	 *
+	 * @throws \Exception
+	 */
+	public function select(array $conditions)
```
Comment (nupplaphil) [inline src/Model/Repository/BaseRepository.php]: The generic selection of entities based on an array of conditions.
It has to be overriden by an explicit repository to set the right "returned" entity-class.

File: src/Model/Repository/BaseRepository.php [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+		$selected = $this->dba->delete(static::$table, [], $conditions);
+
+		$return = [];
+
+		while ($data = $this->dba->fetch($selected)) {
+			if ($prototype === null) {
+				$prototype = new static::$entity($data);
+				$return[] = $prototype;
+			} else {
+				$return[] = static::$entity::createFromPrototype($prototype, $data);
```
Comment (nupplaphil) [inline src/Model/Repository/BaseRepository.php]: This one is for speed :-) .. If we already have a prototype of the current entity, clone it for the rest

File: src/Model/Repository/BaseRepository.php [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+
+		return $data;
+	}
+
+	/**
+	 * @param IStorable[] $entities
+	 *
+	 * @return boolean True, if deleted
+	 */
+	public function delete(IStorable ...$entities)
```
Comment (nupplaphil) [inline src/Model/Repository/BaseRepository.php]: This one is a generic way to delete 1 ... n entities. They just need to implement the `IStorable` interface

File: src/Model/Repository/BaseRepository.php [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+			return false;
+		}
+	}
+
+	/**
+	 * @param IStorable[] $entities
+	 *
+	 * @return boolean True if updated successfully
+	 */
+	public function save(IStorable ...$entities)
```
Comment (nupplaphil) [inline src/Model/Repository/BaseRepository.php]: This is the generic way to save one or more entities.

The save method automatically checks, if an update or insert is needed

**Comment by nupplaphil** [comment]
Interacting with business objects:
```
$intros = DI::repoIntro()->select(['uid' => 1]);
foreach ($intros as $intro) {
    [...]
    $intro->blocked = true;
}
$intros[] = Entity\Introduction::create(['blocked' => false);

// updates all selected intros and inserts the new intro
DI::repoIntro()->save($intros);

foreach ($intros as $intro) {
   [...]
   // since the insert of the last added intro already added the id,  "isStored()" is true :-)
   $this->assertTrue($intros->isStored());
}
```

**Comment by nupplaphil** [comment]
Interacting with dtos:
```php
$instance = DI::factoryInstance()->createInstance();

// exception!
$instance->version = -1;
```

File: src/Model/Entity/BussinessObject/Contact.php [hunk 24]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Friendica\Model\Entity;
+
+use DateTime;
+use Friendica\Model\Entity\Traits\StorableTrait;
+
+/**
+ * @property int      uid
+ * @property DateTime created
```
Comment (nupplaphil) [inline src/Model/Entity/BussinessObject/Contact.php]: One thing I'd love to do is using "real" datetimes for the PHP calls, so we can use the native PHP-functionallty for it :-)

File: src/Model/Entity/BussinessObject/Introduction.php [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace Friendica\Model\Entity;
+
+use DateTime;
+use Friendica\Model\Entity\Traits\StorableTrait;
+
+/**
+ * @property int      uid
+ * @property int      fid
+ * @property Contact  contact
```
Comment (nupplaphil) [inline src/Model/Entity/BussinessObject/Introduction.php]: idea: We could implement a "lazy-loading", means if we want to `get` this attribute, we automatically load the contact for this intro-entity (and log it for optimizing)

File: src/Model/Repository/BaseRepository.php [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+	 *
+	 * @return mixed
+	 *
+	 * @throws \Exception
+	 */
+	public function select(array $conditions)
+	{
+		$prototype = null;
+
+		$selected = $this->dba->delete(static::$table, [], $conditions);
```
Comment (nupplaphil) [inline src/Model/Repository/BaseRepository.php]: idea: we can use defined field-lists dependending on different usages per repository, like
`selectIntrosWithContactData()`
`selectIntrosOnly()`

**Comment by nupplaphil** [comment]
Adding a new entity means doing the following things:
- create a new `Entity` class, which extend `BaseEntity` and adds the owned properties
- create a new `Repository` class ore `Factory` class and set the right entity-class & table

So with 2 classes and a handful code, we would implement a new entity :-) .. At least that's what I think ^^

**Comment by annando** [comment]
I guess that some time in the future I have to take a PHP course. Until then this is simply too high for me.

**Comment by nupplaphil** [comment]
@annando see `entity` as a typesafe array and `repository` as a typesafe, extended `dba` and `factory` as a way to create complex, typesafe arrays.

File: src/Model/Entity/BaseEntity.php (hunk 3 above)
Comment (MrPetovan) [reply src/Model/Entity/BaseEntity.php]: Why don't you make it extend a `DataTransferObject` abstract class that would be what currently is `DataTransferObjectTrait`?

Comment (MrPetovan) [reply src/Model/Entity/BaseEntity.php]: Why not use `DataTransferObjectTrait` as an intermediate class between `BaseEntity` and `Instance`?

For example
```php
abstract class BaseEntity implements ArrayAccess;

abstract class DataTransferEntity extends BaseEntity;
abstract class StorableEntity extends BaseEntity;

class Instance extends DataTransferEntity;
class Contact extends StorableEntity;
```
I personally don't like using traits when you can use class inheritance.

File: src/Model/Entity/BussinessObject/Introduction.php [hunk 27]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace Friendica\Model\Entity;
```
Comment (MrPetovan) [inline src/Model/Entity/BussinessObject/Introduction.php]: Shouldn't it be `Friendica\Model\Entity\BusinessObject` per the folder structure?

File: src/Model/Repository/BaseRepository.php (hunk 22 above)
Comment (MrPetovan) [reply src/Model/Repository/BaseRepository.php]: Why not take an array of entities, or even better, a collection!

File: src/Model/Repository/ContactRepository.php [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+
+	/**
+	 * @param array $conditions
+	 *
+	 * @return Contact[]
+	 * @throws \Exception
+	 */
+	public function select(array $conditions)
+	{
+		return parent::select($conditions);
```
Comment (MrPetovan) [inline src/Model/Repository/ContactRepository.php]: This is what would happen if the method wasn't overloaded, what is the point? Setting the correct return type?

**Review by MrPetovan — COMMENTED** [review]
Thanks for this very detailed prototype. I was planning on working on mine and submitting it as well for review. In your prototype, I feel like we can do away with the static factories. Why not just have the Data Transfer Object have a constructor with the expected data that I had in the `Instance::create` prototype? Wouldn't it simplify the architecture by removing the need for a Factory?

**Comment by annando** [comment]
@nupplaphil my problems are much more basic. I'm already struggling with the naming scheme. I'm struggling with the linked description. I don't know what the "SOLID principle" is. I don't know what "abstractions" are and what they divide from "concrete classes". I don't understand "Dependency Inversion principle". And I guess I will struggle at the explanation of this as well. I'm already stuck with the concept of adding several layers of complexity to solve problems that I don't realize.

I am thinking in database tables and in the IPO model, that's all.

File: src/Model/Entity/BaseEntity.php (hunk 3 above)
Comment (nupplaphil) [reply src/Model/Entity/BaseEntity.php]: I wasn't sure, if they remain just these two traits. Maybe there is particular other functionallity for entities. But you're probably right

File: src/Model/Entity/BussinessObject/Introduction.php (hunk 27 above)
Comment (nupplaphil) [reply src/Model/Entity/BussinessObject/Introduction.php]: Yep

File: src/Model/Repository/BaseRepository.php (hunk 22 above)
Comment (nupplaphil) [reply src/Model/Repository/BaseRepository.php]: I thought so we can pass one entity as well as a list of it. I hadn't tried it, but wouldn't `DI::repo()->delete($entitiyArray);` work as well?

File: src/Model/Repository/ContactRepository.php (hunk 28 above)
Comment (nupplaphil) [reply src/Model/Repository/ContactRepository.php]: Yep.. PHP doesn't support type-overloading. So the base method returns "mixed" and the specific repo has to override it. I didn't find an alternative except explicit creating these methods in each repository again and again.

**Comment by nupplaphil** [comment]
@annando I understand your feedback... There are at least two practical enhancements, which this type of concept allows as:
- avoid uninitialized/wrong typed array access (no more notices :) )
- straight the DB access to fewer points in the code

@mrpetovan
I tried to find just as few ways to create entities as possible in Friendica. And let entities be as few invasive to the code as possible. So we can now dir DTOs and Business objects the same way

With the current approach we could nearly replace arrays with entities and the model methods would remain nearly the same.

And maybe it will be more comfortable for @annando to have a similar concept like tables and rows.

**Comment by nupplaphil** [comment]
Closing this PR in favor of "almost there" design in #8047 :)

**Pull request closed without merging**